crates/bevy_ecs/src/index/mod.rs

```diff
@@ -0,0 +1,468 @@
+        world: UnsafeWorldCell,
+    ) -> bool {
+        let query_valid = <Query<D, (F, With<C>)> as SystemParam>::validate_param(
+            query_state,
+            system_meta,
+            world,
+        );
+        let res_valid =
+            <Res<Index<C>> as SystemParam>::validate_param(res_state, system_meta, world);
+
+        query_valid && res_valid
+    }
+
+    unsafe fn get_param<'world, 'state>(
+        (query_state, res_state): &'state mut Self::State,
+        system_meta: &SystemMeta,
+        world: UnsafeWorldCell<'world>,
+        change_tick: Tick,
+    ) -> Self::Item<'world, 'state> {
+        query_state.validate_world(world.id());
+
+        let index =
+            <Res<Index<C>> as SystemParam>::get_param(res_state, system_meta, world, change_tick);
+
+        QueryByIndex {
+            world,
+            state: None,
+            last_run: system_meta.last_run,
+            this_run: change_tick,
+            index,
+        }
+    }
+}
+
+/// This [`Resource`] is responsible for managing a value-to-[`ComponentId`] mapping, allowing
+/// [`QueryByIndex`] to simply filter by [`ComponentId`] on a standard [`Query`].
+#[derive(Resource)]
+struct Index<C: IndexableComponent> {
+    /// Maps `C` values to a marking ZST component.
+    mapping: HashMap<C, IndexState>,
+    /// Previously registered but currently unused marking ZSTs.
+    /// If a value _was_ tracked in [`mapping`](Index::mapping) but no entity
+    /// has that value anymore, its marker is pushed here for reuse when a _new_
+    /// value for `C` needs to be tracked.
+    ///
+    /// When exhausted, new markers must be registered from a [`World`].
+    spare_markers: Vec<ComponentId>,
+}
+
+/// Internal state for a particular index value within the [`Index`] resource.
+#[derive(Clone, Copy)]
+struct IndexState {
+    /// [`ComponentId`] of this marking ZST
+    component_id: ComponentId,
+    /// A count of how many entities are currently holding this component
+    live_count: usize,
+}
+
+// Rust's derives assume that C must impl Default for this to hold,
+// but that's not true.
+impl<C: IndexableComponent> Default for Index<C> {
+    fn default() -> Self {
+        Self {
+            mapping: Default::default(),
+            spare_markers: Default::default(),
+        }
+    }
+}
+
+impl<C: IndexableComponent> Index<C> {
+    fn track_entity(&mut self, world: &mut World, entity: Entity) {
+        let Some(value) = world.get::<C>(entity) else {
+            return;
+        };
+
+        // Need a marker component for this entity
+        let component_id = match self.mapping.get_mut(value) {
+            // This particular value already has an assigned marker, reuse it.
+            Some(state) => {
+                state.live_count += 1;
+                state.component_id
+            }
+            None => {
+                // Need to clone the index value for later lookups
+                let value = value.clone();
```

Comment:
It doesn't have to be in this PR, but I see a world where the value that is used here is instead of the result of some `IndexableComponent::hash_value()` function, so that we can group components into 'buckets' of values.

That way we could avoid the issue of needing too many ZSTs components for components with float values, as we could for example map all the entities with values between 0.0-1.0 in the same bucket (and the query would then have to do an additional check via a `.iter().filter()` in the `QueryIndex` since there's no guarantee that the entities returned by the index have the exact value that we want).

I'm not super knowledgeable about indexes but basically I think we can re-use a bunch of tricks used by traditional DB indexes